File: plugins/essential/skills/install-agents/scripts/install_agents.py
#!/usr/bin/env python3
"""Discover, preflight, stitch, and install enabled plugin agent templates."""

from __future__ import annotations

import argparse
import json
import os
import re
import shutil
import subprocess
import tempfile
from collections.abc import Iterable
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from stitch_agent import (
    LEAD_AGENT_DIRECTION_ALIAS,
    LEAD_AGENT_DIRECTION_PATH,
    AgentTemplateError,
    load_agent_sources,
    stitch_agent_definition,
    stitch_codex_agent_definition,
)

CACHE_COMPONENT = re.compile(r"^[A-Za-z0-9._-]+$")
# NOTE: Versions additionally admit "+" because SemVer build metadata is
# path-safe and part of the repository's accepted plugin-version contract.
CACHE_VERSION_COMPONENT = re.compile(r"^[A-Za-z0-9._+-]+$")


@dataclass(frozen=True)
class AgentTemplate:
    owner: str
    name: str
    path: Path


def _plugin_templates(owner: str, plugin_root: Path) -> Iterable[AgentTemplate]:
    templates_root = plugin_root / "templates/agents"
    if not templates_root.is_dir():
        return
    resolved_plugin_root = plugin_root.resolve()
    for path in sorted(templates_root.iterdir()):
        if path.is_dir():
            try:
                path.resolve().relative_to(resolved_plugin_root)
            except ValueError as error:
                raise AgentTemplateError(
                    f"template symlink or path escapes plugin root: {path}"
                ) from error
            yield AgentTemplate(owner=owner, name=path.name, path=path)


def _read_plugin_records(harness: str) -> list[dict[str, Any]]:
    command = (
        ["claude", "plugin", "list", "--json"]
        if harness == "claude"
        else ["codex", "plugin", "list", "--json"]
    )
    installed_label = "" if harness == "claude" else "codex "
    try:
        completed = subprocess.run(
            command,
            text=True,
            capture_output=True,
            check=False,
        )
    except OSError as error:
        raise AgentTemplateError(
            f"cannot list installed {installed_label}plugins: {error}"
        ) from error
    if completed.returncode != 0:
        detail = completed.stderr.strip() or completed.stdout.strip()
        raise AgentTemplateError(
            f"cannot list installed {installed_label}plugins: {detail}"
        )
    try:
        payload = json.loads(completed.stdout)
    except json.JSONDecodeError as error:
        raise AgentTemplateError(
            f"invalid JSON from {harness} plugin list: {error}"
        ) from error
    if harness == "claude":
        if not isinstance(payload, list):
            raise AgentTemplateError(
                "claude plugin list --json did not return a list"
            )
        records = payload
    else:
        installed = payload.get("installed") if isinstance(payload, dict) else None
        if not isinstance(installed, list):
            raise AgentTemplateError(
                "codex plugin list --json did not return an installed list"
            )
        records = [
            {
                "id": record.get("pluginId"),
                "enabled": record.get("enabled"),
                "version": record.get("version"),
                "lastUpdated": record.get("lastUpdated"),
            }
            for record in installed
            if isinstance(record, dict)
        ]
    return [record for record in records if isinstance(record, dict)]


def _last_updated(record: dict[str, Any]) -> str:
    value = record.get("lastUpdated")
    return value if isinstance(value, str) else ""


def _codex_cache_plugin_root(
    essential_root: Path,
    record: dict[str, Any],
) -> Path:
    plugin_id = record.get("id")
    version = record.get("version")
    if not isinstance(plugin_id, str) or plugin_id.count("@") != 1:
        raise AgentTemplateError(f"invalid installed Codex plugin id: {plugin_id!r}")
    plugin_name, marketplace = plugin_id.split("@", 1)
    if (
        not isinstance(version, str)
        or plugin_name in {".", ".."}
        or marketplace in {".", ".."}
        or version in {".", ".."}
        or not CACHE_COMPONENT.fullmatch(plugin_name)
        or not CACHE_COMPONENT.fullmatch(marketplace)
        or not CACHE_VERSION_COMPONENT.fullmatch(version)
    ):
        raise AgentTemplateError(
            f"invalid installed Codex plugin cache coordinates: {plugin_id!r} "
            f"version {version!r}"
        )
    cache_root = essential_root.parent.parent.parent
    candidate = cache_root / marketplace / plugin_name / version
    try:
        resolved = candidate.resolve().relative_to(cache_root.resolve())
    except ValueError as error:
        raise AgentTemplateError(
            f"installed Codex plugin cache path escapes cache root: {candidate}"
        ) from error
    installed_root = cache_root / resolved
    if not installed_root.is_dir():
        raise AgentTemplateError(
            f"installed Codex plugin cache root is absent: {installed_root}"
        )
    return installed_root


def _installed_plugin_roots(
    essential_root: Path,
    records: list[dict[str, Any]],
    harness: str,
) -> list[tuple[str, Path]]:
    resolved_essential = essential_root.resolve()
    if harness == "codex":
        if (
            essential_root.parent.name != "essential"
            or essential_root.parent.parent.parent.name != "cache"
        ):
            raise AgentTemplateError(
                f"Codex skill is not loaded from an installed plugin cache: "
                f"{essential_root}"
            )
        cache_marketplace = essential_root.parent.parent.name
        essential_records = [
            record
            for record in records
            if record.get("enabled") is True
            and isinstance(record.get("id"), str)
            and record["id"].count("@") == 1
            and record["id"].split("@", 1)[0] == "essential"
            and record["id"].rsplit("@", 1)[1] == cache_marketplace
            and _codex_cache_plugin_root(essential_root, record).resolve()
            == resolved_essential
        ]
    else:
        essential_records = [
            record
            for record in records
            if isinstance(record.get("installPath"), str)
            and Path(record["installPath"]).resolve() == resolved_essential
            and isinstance(record.get("id"), str)
            and record["id"].count("@") == 1
            and record["id"].split("@", 1)[0] == "essential"
        ]
    if not essential_records:
        raise AgentTemplateError(
            f"essential plugin is absent from {harness} plugin list: {essential_root}"
        )
    if len(essential_records) != 1:
        raise AgentTemplateError(
            f"multiple essential plugin records use install path: {essential_root}"
        )
    essential_record = essential_records[0]
    essential_id = essential_record["id"]
    if "@" not in essential_id:
        raise AgentTemplateError(f"installed plugin id has no marketplace: {essential_id}")
    marketplace = essential_id.rsplit("@", 1)[1]
    if not marketplace:
        raise AgentTemplateError(f"installed plugin id has no marketplace: {essential_id}")
    best_by_id: dict[str, dict[str, Any]] = {}
    for record in records:
        plugin_id = record.get("id")
        install_path = record.get("installPath")
        if (
            record.get("enabled") is not True
            or not isinstance(plugin_id, str)
            or "@" not in plugin_id
            or (harness != "codex" and not isinstance(install_path, str))
        ):
            continue
        record_marketplace = plugin_id.rsplit("@", 1)[1]
        if record_marketplace != marketplace:
            continue
        current = best_by_id.get(plugin_id)
        if current is None or _last_updated(record) > _last_updated(current):
            best_by_id[plugin_id] = record

    roots = sorted(
        (
            (
                plugin_id.rsplit("@", 1)[0],
                (
                    _codex_cache_plugin_root(essential_root, record)
                    if harness == "codex"
                    else Path(record["installPath"])
                ),
            )
            for plugin_id, record in best_by_id.items()
        ),
        key=lambda item: item[0],
    )
    return roots


def discover_agent_templates(
    essential_root: Path,
    plugin_records: list[dict[str, Any]] | None = None,
    harness: str = "claude",
) -> list[AgentTemplate]:
    """Discover source-checkout siblings or enabled same-marketplace installs."""
    essential_root = Path(essential_root)
    if essential_root.parent.name == "plugins":
        roots = [
            (path.name, path)
            for path in sorted(essential_root.parent.iterdir())
            if path.is_dir()
        ]
    else:
        roots = _installed_plugin_roots(
            essential_root,
            (
                plugin_records
                if plugin_records is not None
                else _read_plugin_records(harness)
            ),
            harness,
        )
    return [
        template
        for owner, plugin_root in roots
        for template in _plugin_templates(owner, plugin_root)
    ]


def _preflight(
    templates: list[AgentTemplate],
    harness: str,
    *,
    essential_root: Path,
    reference_root: Path,
    allow_legacy: bool,
) -> list[tuple[str, str]]:
    if not templates:
        raise AgentTemplateError("no agent templates discovered")
    seen: dict[str, AgentTemplate] = {}
    staged: list[tuple[str, str]] = []
    for template in templates:
        sources = load_agent_sources(template.path, allow_legacy=allow_legacy)
        name = sources.metadata["name"]
        previous = seen.get(name)
        if previous is not None:
            raise AgentTemplateError(
                f"duplicate agent name {name!r}: {previous.path} and {template.path}"
            )
        seen[name] = template
        content = (
            stitch_agent_definition(
                template.path,
                essential_root=essential_root,
                reference_root=reference_root,
                allow_legacy=allow_legacy,
            )
            if harness == "claude"
            else stitch_codex_agent_definition(
                template.path,
                essential_root=essential_root,
                reference_root=reference_root,
                allow_legacy=allow_legacy,
            )
        )
        staged.append((name, content))
    return staged


def _replace_file(source: Path, target: Path) -> None:
    target.parent.mkdir(parents=True, exist_ok=True)
    with tempfile.NamedTemporaryFile(
        dir=target.parent,
        prefix=f".{target.name}.",
        suffix=".tmp",
        delete=False,
    ) as temporary_file:
        temporary_path = Path(temporary_file.name)
    try:
        shutil.copy2(source, temporary_path)
        os.replace(temporary_path, target)
    finally:
        temporary_path.unlink(missing_ok=True)


def install_agents(
    essential_root: Path,
    destination: Path,
    plugin_records: list[dict[str, Any]] | None = None,
    harness: str = "claude",
) -> int:
    """Install every discovered template after a complete roster preflight."""
    essential_root = Path(essential_root).resolve()
    templates = discover_agent_templates(essential_root, plugin_records, harness)
    installs_lead_direction = any(
        LEAD_AGENT_DIRECTION_ALIAS
        in (template.path / "base.md").read_text(encoding="utf-8")
        for template in templates
    )
    source_direction = essential_root / LEAD_AGENT_DIRECTION_PATH
    if installs_lead_direction and not source_direction.is_file():
        raise AgentTemplateError(f"missing Essential lead direction: {source_direction}")
    destination = Path(destination)
    installed_essential_root = (destination / ".essential").resolve()
    staged_definitions = _preflight(
        templates,
        harness,
        essential_root=essential_root,
        reference_root=installed_essential_root,
        allow_legacy=essential_root.parent.name != "plugins",
    )
    suffix = ".md" if harness == "claude" else ".toml"
    with tempfile.TemporaryDirectory(prefix=f"{harness}-agents-") as temporary:
        stage = Path(temporary)
        staged_direction = stage / LEAD_AGENT_DIRECTION_PATH
        if installs_lead_direction:
            staged_direction.parent.mkdir(parents=True)
            shutil.copy2(source_direction, staged_direction)
        for name, content in staged_definitions:
            (stage / f"{name}{suffix}").write_text(content, encoding="utf-8")
        destination.mkdir(parents=True, exist_ok=True)
        if installs_lead_direction:
            installed_direction = installed_essential_root / LEAD_AGENT_DIRECTION_PATH
            _replace_file(staged_direction, installed_direction)
            print(f"installed: {installed_direction}")
        for name, _ in staged_definitions:
            target = destination / f"{name}{suffix}"
            _replace_file(stage / target.name, target)
            print(f"installed: {target}")
    print(f"done — installed {len(staged_definitions)} agent(s) into {destination}")
    return len(staged_definitions)


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--plugin-root",
        type=Path,
        default=Path(__file__).resolve().parents[3],
    )
    parser.add_argument(
        "--harness",
        choices=("claude", "codex"),
        default="claude",
    )
    parser.add_argument(
        "--destination",
        type=Path,
    )
    args = parser.parse_args()
    if args.destination is not None:
        destination = args.destination
    elif args.harness == "codex":
        destination = Path(
            os.environ.get("CODEX_HOME", Path.home() / ".codex")
        ) / "agents"
    else:
        destination = Path.home() / ".claude/agents"
    try:
        install_agents(args.plugin_root, destination, harness=args.harness)
    except AgentTemplateError as error:
        parser.error(str(error))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: plugins/essential/skills/install/scripts/install_agents.ts
import {
  existsSync,
  readFileSync,
  readdirSync,
  realpathSync,
  statSync,
} from "node:fs";
import { basename, dirname, relative, resolve, sep } from "node:path";

import {
  AgentTemplateError,
  leadAgentDirectionAlias,
  leadAgentDirectionPath,
  stateSystemsReferencePath,
  loadAgentSources,
  stitchAgentDefinition,
  stitchCodexAgentDefinition,
  stitchGrokAgentDefinition,
} from "./stitch_agent.ts";

import { GrokPluginError, readGrokPlugins } from "../../../scripts/grok.ts";
import {
  defaultAgentDestination,
  installManagedFiles,
} from "../../../scripts/installation.ts";
import {
  canonicalDirectory,
  InstallationError,
} from "../../../scripts/installation/transaction.ts";
import type {
  HarnessName,
  InstallationFile,
  InstallationOptions,
} from "../../../scripts/installation.ts";

type PluginRecord = Record<string, unknown>;
/** discovery and installation settings */
export interface InstallOptions extends InstallationOptions {
  readonly pluginRecords?: readonly PluginRecord[];
  readonly includeMarketplaces?: readonly string[];
}

/** a discovered agent template directory and its owning plugin */
export interface AgentTemplate {
  readonly owner: string;
  readonly name: string;
  readonly path: string;
}
const cacheComponent = /^[A-Za-z0-9._-]+$/;
const cacheVersionComponent = /^[A-Za-z0-9._+-]+$/;
const scriptDirectory = import.meta.dirname;

function inside(root: string, target: string): boolean {
  const path = relative(root, target);
  return (
    path === "" ||
    (path !== ".." && !path.startsWith(`..${sep}`) && !path.startsWith(sep))
  );
}
function isDirectory(path: string): boolean {
  try {
    return statSync(path).isDirectory();
  } catch {
    return false;
  }
}
function resolveExistingAncestors(path: string): string {
  const missing: string[] = [];
  let existing = path;
  while (!existsSync(existing)) {
    const parent = dirname(existing);
    if (parent === existing) return resolve(path);
    missing.unshift(basename(existing));
    existing = parent;
  }
  return resolve(realpathSync(existing), ...missing);
}
function display(value: unknown): string {
  if (typeof value === "string") return `'${value}'`;
  if (value === undefined) return "None";
  return String(value);
}
function pluginTemplates(owner: string, pluginRoot: string): AgentTemplate[] {
  const templatesRoot = resolve(pluginRoot, "agents");
  if (!isDirectory(templatesRoot)) return [];
  const resolvedRoot = realpathSync(pluginRoot);
  return readdirSync(templatesRoot, { withFileTypes: true })
    .filter((entry) => isDirectory(resolve(templatesRoot, entry.name)))
    .sort((left, right) => left.name.localeCompare(right.name))
    .map((entry) => {
      const path = resolve(templatesRoot, entry.name);
      if (!inside(resolvedRoot, realpathSync(path)))
        throw new AgentTemplateError(
          `template symlink or path escapes plugin root: ${path}`,
        );
      return { owner, name: entry.name, path };
    });
}

/**
 * lists installed plugin records by shelling out to the harness CLI
 * @param harness plugin manager whose list command runs
 * @returns normalized plugin records as plain objects
 */
export function readPluginRecords(harness: HarnessName): PluginRecord[] {
  if (harness === "grok") return readGrokPluginRecords();
  const command = [harness, "plugin", "list", "--json"];
  let completed: Bun.SyncSubprocess<"pipe", "pipe">;
  try {
    completed = Bun.spawnSync(command, { stdout: "pipe", stderr: "pipe" });
  } catch (error) {
    throw new AgentTemplateError(
      `cannot list installed ${harness} plugins: ${(error as Error).message}`,
    );
  }
  if (completed.exitCode !== 0) {
    const detail =
      completed.stderr.toString().trim() || completed.stdout.toString().trim();
    throw new AgentTemplateError(
      `cannot list installed ${harness} plugins: ${detail}`,
    );
  }
  let payload: unknown;
  try {
    payload = JSON.parse(completed.stdout.toString());
  } catch (error) {
    throw new AgentTemplateError(
      `invalid JSON from ${harness} plugin list: ${(error as Error).message}`,
    );
  }
  if (harness === "claude") {
    if (!Array.isArray(payload))
      throw new AgentTemplateError(
        "claude plugin list --json did not return a list",
      );
    return payload.filter(
      (record): record is PluginRecord =>
        record !== null && typeof record === "object" && !Array.isArray(record),
    );
  }
  const installed =
    payload !== null && typeof payload === "object" && !Array.isArray(payload)
      ? (payload as PluginRecord).installed
      : undefined;
  if (!Array.isArray(installed))
    throw new AgentTemplateError(
      "codex plugin list --json did not return an installed list",
    );
  return installed
    .filter(
      (record): record is PluginRecord =>
        record !== null && typeof record === "object" && !Array.isArray(record),
    )
    .map((record) => ({
      id: record.pluginId,
      enabled: record.enabled,
      version: record.version,
      lastUpdated: record.lastUpdated,
    }));
}

function lastUpdated(record: PluginRecord): string {
  return typeof record.lastUpdated === "string" ? record.lastUpdated : "";
}

function readGrokPluginRecords(): PluginRecord[] {
  const plugins = readGrokPlugins();
  // installation metadata only supplies trust labels; inspection owns paths and enablement
  let listed: Bun.SyncSubprocess<"pipe", "pipe">;
  try {
    listed = Bun.spawnSync(["grok", "plugin", "list", "--json"], {
      stdout: "pipe",
      stderr: "pipe",
    });
  } catch (error) {
    throw new AgentTemplateError(
      `cannot list Grok plugins: ${(error as Error).message}`,
      { cause: error },
    );
  }
  if (listed.exitCode !== 0) {
    const detail =
      listed.stderr.toString().trim() || listed.stdout.toString().trim();
    throw new AgentTemplateError(
      `cannot list Grok plugins: ${detail || `exit ${listed.exitCode}`}`,
    );
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(listed.stdout.toString());
  } catch (error) {
    throw new AgentTemplateError(
      `invalid JSON from grok plugin list: ${(error as Error).message}`,
      { cause: error },
    );
  }
  if (!Array.isArray(parsed))
    throw new AgentTemplateError(
      "grok plugin list --json did not return a list",
    );
  const metadata = parsed.filter(isRecord);
  return plugins.map((plugin) => {
    const matches = metadata.filter(
      (record) =>
        record.name === plugin.name &&
        [record.path, record.source].some(
          (path) =>
            typeof path === "string" &&
            isDirectory(path) &&
            isDirectory(plugin.path) &&
            realpathSync(path) === realpathSync(plugin.path),
        ),
    );
    const marketplaces = new Set(
      matches
        .map((record) => record.marketplace)
        .filter(
          (marketplace): marketplace is string =>
            typeof marketplace === "string" && marketplace !== "",
        ),
    );
    if (marketplaces.size > 1)
      throw new AgentTemplateError(
        `conflicting Grok marketplace metadata: ${plugin.name}`,
      );
    return {
      id: `${plugin.name}@${[...marketplaces][0] ?? ""}`,
      enabled: plugin.enabled,
      installPath: plugin.path,
    };
  });
}

function grokPluginRoots(
  essentialRoot: string,
  records: readonly PluginRecord[],
  includeMarketplaces: readonly string[],
): Array<readonly [string, string]> {
  const enabled = records.filter((record) => record.enabled === true);
  const essentials = enabled.filter(
    (record) =>
      typeof record.id === "string" &&
      record.id.split("@")[0] === "essential" &&
      typeof record.installPath === "string" &&
      realpathSync(record.installPath) === essentialRoot,
  );
  if (essentials.length !== 1)
    throw new AgentTemplateError(
      `expected one enabled Essential plugin in grok inspect: ${essentialRoot}`,
    );
  const marketplace = String(essentials[0]!.id).split("@")[1] ?? "";
  const trusted = new Set([marketplace, ...includeMarketplaces]);
  for (const included of includeMarketplaces)
    if (!cacheComponent.test(included) || included === "." || included === "..")
      throw new AgentTemplateError(
        `invalid included marketplace name: ${display(included)}`,
      );
  return enabled
    .filter((record) => {
      if (
        typeof record.id !== "string" ||
        record.id.split("@").length !== 2 ||
        typeof record.installPath !== "string"
      )
        throw new AgentTemplateError("invalid enabled Grok plugin record");
      const ownerMarketplace = record.id.split("@")[1]!;
      if (ownerMarketplace !== "") return trusted.has(ownerMarketplace);
      return (
        marketplace === "" &&
        dirname(realpathSync(record.installPath)) === dirname(essentialRoot)
      );
    })
    .map(
      (record) =>
        [String(record.id).split("@")[0]!, String(record.installPath)] as const,
    )
    .sort(([left], [right]) => left.localeCompare(right));
}

function isRecord(value: unknown): value is PluginRecord {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}
/**
 * resolves the installed cache directory of one Codex plugin record
 * @param essentialRoot installed path of the essential plugin
 * @param record one record from the Codex plugin list
 * @returns existing versioned cache directory of the record's plugin
 */
export function codexCachePluginRoot(
  essentialRoot: string,
  record: PluginRecord,
): string {
  const pluginId = record.id;
  const version = record.version;
  if (typeof pluginId !== "string" || pluginId.split("@").length !== 2)
    throw new AgentTemplateError(
      `invalid installed Codex plugin id: ${display(pluginId)}`,
    );
  const [pluginName, marketplace] = pluginId.split("@") as [string, string];
  if (
    typeof version !== "string" ||
    [pluginName, marketplace, version].some((part) =>
      new Set([".", ".."]).has(part),
    ) ||
    !cacheComponent.test(pluginName) ||
    !cacheComponent.test(marketplace) ||
    !cacheVersionComponent.test(version)
  )
    throw new AgentTemplateError(
      `invalid installed Codex plugin cache coordinates: ${display(pluginId)} version ${display(version)}`,
    );
  const cacheRoot = resolve(essentialRoot, "../../..");
  const candidate = resolve(cacheRoot, marketplace, pluginName, version);
  if (!inside(realpathSync(cacheRoot), resolveExistingAncestors(candidate)))
    throw new AgentTemplateError(
      `installed Codex plugin cache path escapes cache root: ${candidate}`,
    );
  if (!isDirectory(candidate))
    throw new AgentTemplateError(
      `installed Codex plugin cache root is absent: ${candidate}`,
    );
  return candidate;
}

/**
 * pairs every enabled plugin owner with its best installed root directory
 * @param essentialRoot installed path of the essential plugin
 * @param records records from the harness plugin list
 * @param harness plugin manager the records came from
 * @param includeMarketplaces extra trusted marketplaces to include
 * @returns owner-to-root tuples sorted by owner name
 */
export function installedPluginRoots(
  essentialRoot: string,
  records: readonly PluginRecord[],
  harness: HarnessName,
  includeMarketplaces: readonly string[] = [],
): Array<readonly [string, string]> {
  const resolvedEssential = realpathSync(essentialRoot);
  if (harness === "grok")
    return grokPluginRoots(resolvedEssential, records, includeMarketplaces);
  let essentialRecords: readonly PluginRecord[];
  if (harness === "codex") {
    if (
      basename(dirname(essentialRoot)) !== "essential" ||
      basename(resolve(essentialRoot, "../../..")) !== "cache"
    )
      throw new AgentTemplateError(
        `Codex skill is not loaded from an installed plugin cache: ${essentialRoot}`,
      );
    const marketplace = basename(resolve(essentialRoot, "../.."));
    essentialRecords = records.filter(
      (record) =>
        record.enabled === true &&
        typeof record.id === "string" &&
        record.id.split("@").length === 2 &&
        record.id.split("@")[0] === "essential" &&
        record.id.split("@")[1] === marketplace &&
        realpathSync(codexCachePluginRoot(essentialRoot, record)) ===
          resolvedEssential,
    );
  } else
    essentialRecords = records.filter(
      (record) =>
        typeof record.installPath === "string" &&
        realpathSync(record.installPath) === resolvedEssential &&
        typeof record.id === "string" &&
        record.id.split("@").length === 2 &&
        record.id.split("@")[0] === "essential",
    );
  if (essentialRecords.length === 0)
    throw new AgentTemplateError(
      `essential plugin is absent from ${harness} plugin list: ${essentialRoot}`,
    );
  if (essentialRecords.length !== 1)
    throw new AgentTemplateError(
      `multiple essential plugin records use install path: ${essentialRoot}`,
    );
  const essentialId = String(essentialRecords[0]!.id);
  const marketplace = essentialId.split("@").at(-1) ?? "";
  if (marketplace === "")
    throw new AgentTemplateError(
      `installed plugin id has no marketplace: ${essentialId}`,
    );
  const marketplaces = new Set([marketplace]);
  for (const included of includeMarketplaces) {
    if (new Set([".", ".."]).has(included) || !cacheComponent.test(included))
      throw new AgentTemplateError(
        `invalid included marketplace name: ${display(included)}`,
      );
    marketplaces.add(included);
  }
  const best = new Map<string, PluginRecord>();
  for (const record of records) {
    const id = record.id;
    if (
      record.enabled !== true ||
      typeof id !== "string" ||
      !id.includes("@") ||
      (harness !== "codex" && typeof record.installPath !== "string") ||
      !marketplaces.has(id.split("@").at(-1) ?? "")
    )
      continue;
    const current = best.get(id);
    if (current === undefined || lastUpdated(record) > lastUpdated(current))
      best.set(id, record);
  }
  return [...best]
    .map(
      ([id, record]) =>
        [
          id.split("@")[0]!,
          harness === "codex"
            ? codexCachePluginRoot(essentialRoot, record)
            : String(record.installPath),
        ] as const,
    )
    .sort(([left], [right]) => left.localeCompare(right));
}

/**
 * discovers agent template directories across source checkout or installed roots
 * @param essentialRoot path of the essential plugin
 * @param options pluginRecords supply harness records directly; harness selects
 *   the plugin manager; includeMarketplaces adds trusted marketplaces
 * @returns templates sorted by owner then directory name
 */
export function discoverAgentTemplates(
  essentialRoot: string,
  options: {
    readonly pluginRecords?: readonly PluginRecord[];
    readonly harness?: HarnessName;
    readonly includeMarketplaces?: readonly string[];
  } = {},
): AgentTemplate[] {
  const harness = options.harness ?? "claude";
  const parent = dirname(essentialRoot);
  const roots: Array<readonly [string, string]> =
    harness !== "grok" && basename(parent) === "plugins"
      ? readdirSync(parent, { withFileTypes: true })
          .filter((entry) => entry.isDirectory())
          .sort((left, right) => left.name.localeCompare(right.name))
          .map((entry) => [entry.name, resolve(parent, entry.name)] as const)
      : installedPluginRoots(
          essentialRoot,
          options.pluginRecords ?? readPluginRecords(harness),
          harness,
          options.includeMarketplaces,
        );
  return roots.flatMap(([owner, root]) => pluginTemplates(owner, root));
}

function preflight(
  templates: readonly AgentTemplate[],
  harness: HarnessName,
  options: {
    readonly essentialRoot: string;
    readonly referenceRoot: string;
    readonly allowLegacy: boolean;
  },
): Array<readonly [string, string]> {
  if (templates.length === 0 && harness !== "grok")
    throw new AgentTemplateError("no agent templates discovered");
  const seen = new Map<string, AgentTemplate>();
  return templates.map((template) => {
    const sources = loadAgentSources(template.path, {
      allowLegacy: options.allowLegacy,
    });
    const name = String(sources.metadata.name);
    const previous = seen.get(name);
    if (previous !== undefined)
      throw new AgentTemplateError(
        `duplicate agent name ${display(name)}: ${previous.path} and ${template.path}`,
      );
    seen.set(name, template);
    const stitchOptions = {
      essentialRoot: options.essentialRoot,
      referenceRoot: options.referenceRoot,
      allowLegacy: options.allowLegacy,
    };
    return [
      name,
      harness === "claude"
        ? stitchAgentDefinition(template.path, stitchOptions)
        : harness === "codex"
          ? stitchCodexAgentDefinition(template.path, stitchOptions)
          : stitchGrokAgentDefinition(template.path, stitchOptions),
    ] as const;
  });
}

/**
 * stages every discovered template and installs it atomically into the destination
 * @param essentialRoot path of the essential plugin
 * @param destination directory receiving the stitched agent files
 * @param options pluginRecords, harness, and includeMarketplaces feed discovery;
 *   stdout overrides progress reporting
 * @returns number of agents installed
 */
export function installAgents(
  essentialRoot: string,
  destination: string,
  options: InstallOptions = {},
): number {
  const root = realpathSync(essentialRoot);
  const harness = options.harness ?? "claude";
  const templates = discoverAgentTemplates(root, {
    pluginRecords: options.pluginRecords,
    harness,
    includeMarketplaces: options.includeMarketplaces,
  });
  const installsDirection = templates.some((template) =>
    readFileSync(resolve(template.path, "base.md"), "utf8").includes(
      leadAgentDirectionAlias,
    ),
  );
  const sourceDirection = resolve(root, leadAgentDirectionPath);
  if (installsDirection && !existsSync(sourceDirection))
    throw new AgentTemplateError(
      `missing Essential lead direction: ${sourceDirection}`,
    );
  const sourceStateSystems = resolve(root, stateSystemsReferencePath);
  if (templates.length > 0 && !existsSync(sourceStateSystems))
    throw new AgentTemplateError(
      `missing Essential state-system authority: ${sourceStateSystems}`,
    );
  const canonicalDestination = canonicalDirectory(destination);
  const installedEssential = resolve(canonicalDestination, ".essential");
  const staged = preflight(templates, harness, {
    essentialRoot: root,
    referenceRoot: installedEssential,
    allowLegacy: basename(dirname(root)) !== "plugins",
  });
  const suffix = harness === "codex" ? ".toml" : ".md";
  const write =
    options.stdout ??
    ((text: string): void => {
      process.stdout.write(text);
    });
  const files: InstallationFile[] = staged.map(([name, content]) => ({
    path: `${name}${suffix}`,
    content,
    kind: "agent",
  }));
  if (staged.length > 0)
    files.unshift({
      path: `.essential/${stateSystemsReferencePath}`,
      content: readFileSync(sourceStateSystems),
      kind: "support",
    });
  if (installsDirection)
    files.push({
      path: `.essential/${leadAgentDirectionPath}`,
      content: readFileSync(sourceDirection),
      kind: "support",
    });
  installManagedFiles(
    canonicalDestination,
    files,
    { ...options, harness, stdout: write },
    resolve(root, "directions/GROK.md"),
  );
  write(`done — installed ${staged.length} agent(s) into ${destination}\n`);
  return staged.length;
}

const program = basename(import.meta.url);
const usage = `usage: ${program} [-h] [--plugin-root PLUGIN_ROOT]\n${" ".repeat(program.length + 7)}[--harness {claude,codex,grok}]\n${" ".repeat(program.length + 7)}[--destination DESTINATION]\n${" ".repeat(program.length + 7)}[--include-marketplace INCLUDE_MARKETPLACE]`;
const help = `${usage}\n\nDiscover, preflight, stitch, and install enabled plugin agent templates.\n\noptions:\n  -h, --help            show this help message and exit\n  --plugin-root PLUGIN_ROOT\n  --harness {claude,codex,grok}\n  --destination DESTINATION\n  --include-marketplace INCLUDE_MARKETPLACE\n                        also discover enabled agent templates from this\n                        trusted marketplace\n`;
function cliError(message: string): never {
  process.stderr.write(`${usage}\n${program}: error: ${message}\n`);
  process.exit(2);
}
/**
 * parses installer flags and drives installAgents for one harness
 * @param argv arguments following the script name
 * @returns process exit code: 0 success, 2 usage error
 */
export function main(argv = process.argv.slice(2)): number {
  let pluginRoot = resolve(scriptDirectory, "../../..");
  let harness: HarnessName = "claude";
  let destination: string | undefined;
  const includeMarketplaces: string[] = [];
  for (let index = 0; index < argv.length; index += 1) {
    const argument = argv[index]!;
    const value = (): string => {
      const next = argv[index + 1];
      if (next === undefined || next.startsWith("-"))
        cliError(`argument ${argument}: expected one argument`);
      index += 1;
      return next;
    };
    if (argument === "-h" || argument === "--help") {
      process.stdout.write(help);
      return 0;
    }
    if (argument === "--plugin-root") pluginRoot = value();
    else if (argument.startsWith("--plugin-root="))
      pluginRoot = argument.slice("--plugin-root=".length);
    else if (argument === "--harness") {
      const selected = value();
      if (selected !== "claude" && selected !== "codex" && selected !== "grok")
        cliError(
          `argument --harness: invalid choice: '${selected}' (choose from 'claude', 'codex', 'grok')`,
        );
      harness = selected;
    } else if (argument.startsWith("--harness=")) {
      const selected = argument.slice("--harness=".length);
      if (selected !== "claude" && selected !== "codex" && selected !== "grok")
        cliError(
          `argument --harness: invalid choice: '${selected}' (choose from 'claude', 'codex', 'grok')`,
        );
      harness = selected;
    } else if (argument === "--destination") destination = value();
    else if (argument.startsWith("--destination="))
      destination = argument.slice("--destination=".length);
    else if (argument === "--include-marketplace")
      includeMarketplaces.push(value());
    else if (argument.startsWith("--include-marketplace="))
      includeMarketplaces.push(argument.slice("--include-marketplace=".length));
    else cliError(`unrecognized arguments: ${argument}`);
  }
  destination ??= defaultAgentDestination(harness);
  try {
    installAgents(pluginRoot, destination, { harness, includeMarketplaces });
    return 0;
  } catch (error) {
    if (
      error instanceof AgentTemplateError ||
      error instanceof InstallationError ||
      error instanceof GrokPluginError
    )
      cliError(error.message);
    throw error;
  }
}

if (import.meta.main) process.exit(main());
